default/python/universe_generation/universe_generator.py

```diff
@@ -69,6 +72,97 @@ def error_report():
     return error_list
 
 
+def read_starting_unlocks(starting_era):
+    starting_unlocks = {
+        "buildings": [],
+        "items": [],
+        "fleets": [],
+    }
+
+    unlock_dir = os.path.join(fo.get_resource_dir(), "scripting",
+                              "starting_unlocks")
+    if starting_era == fo.startingEra.prewarp:
+        buildings_inf = os.path.join(unlock_dir, "buildings-prewarp.inf")
+        items_inf = os.path.join(unlock_dir, "items-prewarp.inf")
+    else:
+        buildings_inf = os.path.join(unlock_dir, "buildings.inf")
+        items_inf = os.path.join(unlock_dir, "items.inf")
+        fleets_inf = os.path.join(unlock_dir, "fleets.inf")
+
+    comment_expr = re.compile(r'\s*(?://|$)')
+    try:
+        f = open(buildings_inf)
+    except IOError:
+        print >> sys.stderr, "Cannot read", buildings_inf
+    else:
+        expr = re.compile(r'.*Building\s+name\s+=\s+"(.+)"')
+        for line in f:
+            if comment_expr.match(line):
+                continue
+            match = expr.match(line)
+            if match is None:
+                print >> sys.stderr, "buildings.inf parse error:", line
+            else:
+                starting_unlocks["buildings"].append(match.group(1))
+        f.close()
```

Comment:
Implemented try: ... finally: protection in commit 6e46fd75.

I dislike the idiom "with open" because it implies that the parsing code will be running inside the same try: clause as the open(), so the error handling cannot distinguish exceptions from the open() from exceptions from the parser.